### PR TITLE
75277 update requirement definition

### PR DESCRIPTION
**Pull request by bratli**
Endpoint for updating requirement definition.
-The endpoint should do all updates that are possible in the view that can be seen in figma

NB: Deleting/voiding fields needs to be further discussed - maybe wait until delete functionality is decided on as well.  A new PBI has been created for voiding/deleting fields

### CODE REVIEW

File: src/Equinor.Procosys.Preservation.Command/RequirementTypeCommands/CreateRequirementDefinition/CreateRequirementDefinitionCommandHandler.cs [hunk 1]
```diff
@@ -1,4 +1,5 @@
-﻿using System.Linq;
+﻿using System.Collections.Generic;
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/RequirementTypeCommands/CreateRequirementDefinition/CreateRequirementDefinitionCommandHandler.cs:1]: remove

File: src/Equinor.Procosys.Preservation.Command/RequirementTypeCommands/UpdateRequirementDefinition/UpdateFieldsForCommand.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        }
+
+        public int Id { get; }
+        public string Label { get; }
+        public string Unit { get; private set; }
+        public bool? ShowPrevious { get; }
+        public int SortKey { get; }
+        public FieldType FieldType { get; }
+        public string RowVersion { get;  }
+        public bool? Delete { get; }
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/RequirementTypeCommands/UpdateRequirementDefinition/UpdateFieldsForCommand.cs:34]: replace with IsVoided

File: src/Equinor.Procosys.Preservation.Command/RequirementTypeCommands/UpdateRequirementDefinition/UpdateRequirementDefinitionCommandHandler.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+                requirementType.RequirementDefinitions.Single(rd => rd.Id == request.RequirementDefinitionId);
+
+            requirementDefinition.Title = request.Title;
+            requirementDefinition.SortKey = request.SortKey;
+            requirementDefinition.Usage = request.Usage;
+            requirementDefinition.DefaultIntervalWeeks = request.DefaultIntervalWeeks;
+
+            foreach (var f in request.UpdateFields)
+            {
+                var fieldToUpdate = requirementDefinition.Fields.Single(field => field.Id == f.Id);
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/RequirementTypeCommands/UpdateRequirementDefinition/UpdateRequirementDefinitionCommandHandler.cs:40]: Handle IsVoided

File: src/Equinor.Procosys.Preservation.Command/RequirementTypeCommands/UpdateRequirementDefinition/UpdateRequirementDefinitionCommandValidator.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            RuleForEach(command => command.UpdateFields)
+                .MustAsync((command, field, __, token) => BeAnExistingFieldToUpdate(field.Id, token))
+                .WithMessage((_, field) => $"Field doesn't exist'! Field={field.Id}")
+                .MustAsync((command, field, __, token) => BeSameFieldTypeOnExistingFieldsAsync(field, token))
+                .WithMessage((_, field) => $"Cannot change field type on existing fields! Field={field.Id}");
+
+            async Task<bool> BeAnExistingRequirementTypeAsync(int requirementTypeId, CancellationToken token)
+                => await requirementTypeValidator.ExistsAsync(requirementTypeId, token);
+            async Task<bool> BeAnExistingRequirementDefinitionAsync(int requirementDefinitionId, CancellationToken token)
+                => await requirementDefinitionValidator.ExistsAsync(requirementDefinitionId, token);
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/RequirementTypeCommands/UpdateRequirementDefinition/UpdateRequirementDefinitionCommandValidator.cs:43]: Must check that definition id exists within the type

File: src/Equinor.Procosys.Preservation.Command/Validators/FieldValidators/FieldValidator.cs [hunk 5]
```diff
@@ -1,6 +1,8 @@
-﻿using System.Linq;
+﻿using System.Collections.Generic;
+using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
+using Equinor.Procosys.Preservation.Command.RequirementTypeCommands.UpdateRequirementDefinition;
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/Validators/FieldValidators/FieldValidator.cs:5]: remove, use simple params in method

File: src/Equinor.Procosys.Preservation.Command/Validators/FieldValidators/FieldValidator.cs [hunk 6]
```diff
@@ -41,5 +43,14 @@ public async Task<bool> IsValidForAttachmentAsync(int fieldId, CancellationToken
                 select f).SingleOrDefaultAsync(token);
             return field != null && (field.FieldType == FieldType.Attachment);
         }
+
+        public async Task<bool> FieldTypeIsNotChanged(UpdateFieldsForCommand fieldToUpdate, CancellationToken token)
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/Validators/FieldValidators/FieldValidator.cs:47]: rename to VerifyFieldTypeAsync

File: src/Equinor.Procosys.Preservation.Command/Validators/FieldValidators/IFieldValidator.cs [hunk 7]
```diff
@@ -9,5 +10,6 @@ public interface IFieldValidator
         Task<bool> IsVoidedAsync(int fieldId, CancellationToken token);
         Task<bool> IsValidForRecordingAsync(int fieldId, CancellationToken token);
         Task<bool> IsValidForAttachmentAsync(int fieldId, CancellationToken token);
+        Task<bool> FieldTypeIsNotChanged(UpdateFieldsForCommand fieldToUpdate, CancellationToken token);
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/Validators/FieldValidators/IFieldValidator.cs:13]: rename to VerifyFieldTypeAsync

Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/Validators/FieldValidators/IFieldValidator.cs:13]: Add unit test

File: src/Equinor.Procosys.Preservation.WebApi/Controllers/RequirementTypes/RequirementTypesController.cs [hunk 8]
```diff
@@ -176,5 +177,49 @@ public class RequirementTypesController : ControllerBase
             var result = await _mediator.Send(command);
             return Ok(result);
         }
+
+        [Authorize(Roles = Permissions.LIBRARY_PRESERVATION_VOIDUNVOID)]
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.WebApi/Controllers/RequirementTypes/RequirementTypesController.cs:181]: incorrect role

File: src/Equinor.Procosys.Preservation.Command/Validators/RequirementDefinitionValidators/IRequirementDefinitionValidator.cs [hunk 9]
```diff
@@ -12,7 +12,16 @@ public interface IRequirementDefinitionValidator
         Task<bool> UsageCoversBothForSupplierAndOtherAsync(List<int> requirementDefinitionIds, CancellationToken token);
         Task<bool> UsageCoversForOtherThanSuppliersAsync(List<int> requirementDefinitionIds, CancellationToken token);
         Task<bool> HasAnyForSupplierOnlyUsageAsync(List<int> requirementDefinitionIds, CancellationToken token);
-        Task<bool> IsNotUniqueTitleOnRequirementTypeAsync(int requirementTypeId, string reqDefTitle, IEnumerable<Field> fields,
+        Task<bool> IsNotUniqueTitleOnRequirementTypeAsync(
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/Validators/RequirementDefinitionValidators/IRequirementDefinitionValidator.cs:15]: move to IRequirementTypeValidator. Rename to RequirementDefinitionExistsAsync

File: src/Equinor.Procosys.Preservation.Command/Validators/RequirementDefinitionValidators/IRequirementDefinitionValidator.cs [hunk 10]
```diff
@@ -12,7 +12,16 @@ public interface IRequirementDefinitionValidator
...
+            int requirementTypeId, 
+            string reqDefTitle,
+            IList<FieldType> fieldTypes,
+            CancellationToken token);
+        Task<bool> IsNotUniqueUpdatedTitleOnRequirementTypeAsync(
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/Validators/RequirementDefinitionValidators/IRequirementDefinitionValidator.cs:20]: move to IRequirementTypeValidator. Rename to RequirementDefinitionExistsAsync

File: src/Equinor.Procosys.Preservation.WebApi/Controllers/RequirementTypes/UpdateFieldDto.cs [hunk 11]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Equinor.Procosys.Preservation.WebApi.Controllers.RequirementTypes
+{
+    public class UpdateFieldDto : FieldDto
+    {
+        public int Id { get; set; }
+        public string RowVersion { get; set; }
+        public bool Delete { get; set; }
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.WebApi/Controllers/RequirementTypes/UpdateFieldDto.cs:7]: Void

File: src/Equinor.Procosys.Preservation.WebApi/Controllers/RequirementTypes/UpdateFieldDtoValidator.cs [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+                .MaximumLength(Field.UnitLengthMax);
+
+            RuleFor(x => x.SortKey)
+                .NotNull()
+                .WithMessage("Sort key for field cannot be null");
+
+            RuleFor(x => x.FieldType)
+                .NotNull();
+
+            RuleFor(x => x.SortKey)
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.WebApi/Controllers/RequirementTypes/UpdateFieldDtoValidator.cs:28]: use validator at line 21

File: src/Equinor.Procosys.Preservation.WebApi/Controllers/RequirementTypes/UpdateRequirementDefinitionDtoValidator.cs [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+        public UpdateRequirementDefinitionDtoValidator()
+        {
+            RuleFor(x => x).NotNull();
+
+            RuleFor(x => x.Title)
+                .MaximumLength(RequirementDefinition.TitleLengthMax);
+
+            RuleFor(x => x.SortKey)
+                .NotNull()
+                .Must(MustBePositive)
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.WebApi/Controllers/RequirementTypes/UpdateRequirementDefinitionDtoValidator.cs:18]: remove

File: src/Equinor.Procosys.Preservation.WebApi/Controllers/RequirementTypes/UpdateRequirementDefinitionDtoValidator.cs [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+                .WithMessage("Sort key cannot be null");
+
+            RuleFor(x => x.SortKey)
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.WebApi/Controllers/RequirementTypes/UpdateRequirementDefinitionDtoValidator.cs:21]: merge with above

File: src/tests/Equinor.Procosys.Preservation.Command.Tests/MiscCommands/Clone/RequirementTypeRepository.cs [hunk 15]
```diff
@@ -28,5 +28,6 @@ public void Save()
 
         public Task<RequirementDefinition> GetRequirementDefinitionByIdAsync(int requirementDefinitionId) => throw new NotImplementedException();
         public Task<List<RequirementDefinition>> GetRequirementDefinitionsByIdsAsync(IList<int> requirementDefinitionIds) => throw new NotImplementedException();
+        public void RemoveField(Field field) => throw new NotImplementedException();
```
Comment (janid1967) [inline src/tests/Equinor.Procosys.Preservation.Command.Tests/MiscCommands/Clone/RequirementTypeRepository.cs:31]: remove

File: src/tests/Equinor.Procosys.Preservation.Command.Tests/Validators/RequirementDefinitionValidatorTests.cs [hunk 16]
```diff
@@ -323,5 +325,29 @@ public async Task UsageCoversBothForSupplierAndOtherAsync_NoRequirements_Returns
                 Assert.IsFalse(result);
             }
         }
+
```
Comment (janid1967) [inline src/tests/Equinor.Procosys.Preservation.Command.Tests/Validators/RequirementDefinitionValidatorTests.cs:328]: move to TypeValidator

File: src/Equinor.Procosys.Preservation.WebApi/Controllers/RequirementTypes/CreateRequirementDefinitionDtoValidator.cs [hunk 17]
```diff
@@ -16,14 +16,16 @@ public CreateRequirementDefinitionDtoValidator()
 
             RuleFor(x => x.SortKey)
                 .NotNull()
-                .WithMessage("Sort key cannot be null");
+                .WithMessage("Sort key cannot be null")
+                .Must(MustBePositive)
```
Comment (llykkelarsen) [inline src/Equinor.Procosys.Preservation.WebApi/Controllers/RequirementTypes/CreateRequirementDefinitionDtoValidator.cs:20]: Must(BePositive)? 2 x must

**Pull request merged**